File: hybrid/SampleApps/SmartStorez/SmartStorez/www/SFSmartstorePlugin.js
/*
 * Copyright (c) 2011, salesforce.com, inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided
 * that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this list of conditions and the
 * following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and
 * the following disclaimer in the documentation and/or other materials provided with the distribution.
 *
 * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to endorse or
 * promote products derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */



if (!PhoneGap.hasResource("smartstore")) {

PhoneGap.addResource("smartstore");

/**
 *  SmartStoreError.
 *  An error code assigned by an implementation when an error has occurred
 */
var SmartStoreError = function () {
    this.code = null;
};

/**
 * Error codes
 */
SmartStoreError.UNKNOWN_ERROR = 0;
SmartStoreError.INVALID_ARGUMENT_ERROR = 1;
SmartStoreError.TIMEOUT_ERROR = 2;
SmartStoreError.PENDING_OPERATION_ERROR = 3;
SmartStoreError.IO_ERROR = 4;
SmartStoreError.NOT_SUPPORTED_ERROR = 5;
SmartStoreError.PERMISSION_DENIED_ERROR = 20;
 
/**
 * IndexSpec
 */
var SoupIndexSpec = function (path,type) {
    this.path = path;
    this.type = type;
};

/**
 * QuerySpec
 */
var SoupQuerySpec = function (path,matchKey) {
    //path for the original IndexSpec you wish to use for search: may be a compound path eg Account.Owner.Name
    this.indexPath = path;
    //exact match key 
    this.matchKey = matchKey;
    
    //the value at which query results may begin
    this.beginKey = null;
    //the value at which query results may end
    this.endKey = null;

    //"ascending" or "descending" : optional
    this.order = "ascending";

    //the number of entries to copy from native to javascript per each cursor page
    this.pageSize = 10;
};

/**
 * Cursor
 */
var PagedSoupCursor = function () {
    //the soup name from which this cursor was generated
    this.soupName = null;
    //a unique identifier for this cursor, used by plugin
    this.cursorId = null;
    //the query spec that generated this cursor
    this.querySpec = null;
    //the maximum number of entries returned per page 
    this.pageSize = 0;
    //the total number of pages of results available
    this.totalPages = 0;
    //the current page index among all the pages available
    this.currentPageIndex = 0;

    //the list of current page entries, ordered as requested in the querySpec
    this.currentPageOrderedEntries = null;
};

/**
 * Represents a group of Contacts.
 */
var SmartStore = function () {
    logToConsole("new SmartStore");
};
 
// ====== Soup manipulation ======

SmartStore.prototype.registerSoup = function (soupName,indexSpecs,successCB,errorCB) {
    logToConsole("SmartStore.registerSoup: '" + soupName + "' indexSpecs: " + indexSpecs);
    
    PhoneGap.exec(successCB, errorCB, 
                  "com.salesforce.smartstore",
                  "pgRegisterSoup",
                  [{"soupName":soupName,"indexes":indexSpecs}]
                  );                  
};

SmartStore.prototype.removeSoup = function (soupName,successCB,errorCB) {
    logToConsole("SmartStore.removeSoup: " + soupName );
    
    PhoneGap.exec(successCB, errorCB, 
                  "com.salesforce.smartstore",
                  "pgRemoveSoup",
                  [{"soupName":soupName}]
                  );                  
};

SmartStore.prototype.querySoup = function (soupName,querySpec,successCB,errorCB) {
    logToConsole("SmartStore.querySoup: '" + soupName + "' indexPath: " + querySpec.indexPath);
    
    PhoneGap.exec(successCB, errorCB, 
                  "com.salesforce.smartstore",
                  "pgQuerySoup",
                  [{"soupName":soupName,"querySpec":querySpec}]
                  );
};

SmartStore.prototype.retrieveSoupEntry = function (soupName,soupEntryId,successCB,errorCB) {
    logToConsole("SmartStore.retrieveSoupEntry: '" + soupName + "' soupEntryId: " + soupEntryId);
    
    PhoneGap.exec(successCB, errorCB, 
                  "com.salesforce.smartstore",
                  "pgRetrieveSoupEntry",
                  [{"soupName":soupName,"soupEntryId":soupEntryId}]
                  );
};

SmartStore.prototype.upsertSoupEntries = function (soupName,entries,successCB,errorCB) {
    logToConsole("SmartStore.upsertSoupEntries: '" + soupName + "' entries: " + entries.length);

    PhoneGap.exec(successCB, errorCB, 
                  "com.salesforce.smartstore",
                  "pgUpsertSoupEntries",
                  [{"soupName":soupName,"entries":entries}]
                  );
};

SmartStore.prototype.removeFromSoup = function (soupName,entryIds,successCB,errorCB) {
    logToConsole("SmartStore.removeFromSoup: '" + soupName + "' entryIds: " + entryIds.length);

    PhoneGap.exec(successCB, errorCB, 
                  "com.salesforce.smartstore",
                  "pgRemoveFromSoup",
                  [{"soupName":soupName,"entryIds":entryIds}]
                  );
};

//====== Cursor manipulation ======
    
SmartStore.prototype.moveCursorToPageIndex = function (cursor,newPageIndex,successCB,errorCB) {
    logToConsole("moveCursorToPageIndex " + newPageIndex);

    PhoneGap.exec(successCB, errorCB, 
    "com.salesforce.smartstore",
    "pgMoveCursorToPageIndex",
    [{"cursorId":cursor.cursorId, "index":newPageIndex}]
    );
};

SmartStore.prototype.moveCursorToNextPage = function (cursor,successCB,errorCB) {
    var newPageIndex = cursor.currentPageIndex + 1;
    if (newPageIndex >= cursor.totalPages) {
        return;//TODO callback with error?
    }

    this.moveCursorToPageIndex(cursor,newPageIndex,successCB,errorCB);
};

SmartStore.prototype.moveCursorToPreviousPage = function (cursor,successCB,errorCB) {
    var newPageIndex = cursor.currentPageIndex - 1;
    if (newPageIndex < 0) {
        return;//TODO callback with error?
    }

    this.moveCursorToPageIndex(cursor,newPageIndex,successCB,errorCB);
};

//======Plugin creation / installation ======
    

    
PhoneGap.addConstructor(function () {
        logToConsole("SmartStore pre-install");
         if (typeof navigator.smartstore === 'undefined') {
             logToConsole("SmartStore.install");
             navigator.smartstore = new SmartStore();
         }
});

}


    

File: hybrid/SalesforceHybridSDK/SalesforceHybridSDK/Plugins/SFSmartStore/SFSmartStorePlugin.m
/*
 Copyright (c) 2012, salesforce.com, inc. All rights reserved.
 Author: Todd Stellanova
 
 Redistribution and use of this software in source and binary forms, with or without modification,
 are permitted provided that the following conditions are met:
 * Redistributions of source code must retain the above copyright notice, this list of conditions
 and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list of
 conditions and the following disclaimer in the documentation and/or other materials provided
 with the distribution.
 * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
 endorse or promote products derived from this software without specific prior written
 permission of salesforce.com, inc.
 
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "SFSmartStorePlugin.h"

#import "NSDictionary+SFAdditions.h"

#import "SFContainerAppDelegate.h"
#import "SFSoupCursor.h"
#import "SFSmartStore.h"
#import "SFHybridViewController.h"
#import <Cordova/CDVPlugin.h>
#import <Cordova/CDVPluginResult.h>

//NOTE: must match value in Cordova.plist file
NSString * const kSmartStorePluginIdentifier = @"com.salesforce.smartstore";

// Private constants

NSString * const kSoupNameArg         = @"soupName";
NSString * const kEntryIdsArg         = @"entryIds";
NSString * const kCursorIdArg         = @"cursorId";
NSString * const kIndexArg            = @"index";
NSString * const kIndexesArg          = @"indexes";
NSString * const kQuerySpecArg        = @"querySpec";
NSString * const kEntriesArg          = @"entries";
NSString * const kExternalIdPathArg   = @"externalIdPath";



@interface SFSmartStorePlugin() 

- (void)closeCursorWithId:(NSString *)cursorId;

@end





@implementation SFSmartStorePlugin


@synthesize cursorCache = _cursorCache;
@synthesize store = _store;


+ (void)resetSharedStore {
    SFContainerAppDelegate *myApp = (SFContainerAppDelegate*)[[UIApplication sharedApplication] delegate];
    SFSmartStorePlugin *myInstance = (SFSmartStorePlugin*)[myApp.viewController.commandDelegate getCommandInstance:kSmartStorePluginIdentifier];
    [[myInstance cursorCache] removeAllObjects];
    myInstance.store = nil; 
    myInstance.store = [SFSmartStore sharedStoreWithName:kDefaultSmartStoreName];
}

- (CDVPlugin*) initWithWebView:(UIWebView*)theWebView 
{
    self = [super initWithWebView:theWebView];
    
    if (nil != self)  {
        NSLog(@"SFSmartStorePlugin initWithWebView");
        _appDelegate = (SFContainerAppDelegate *)[self appDelegate];
        _cursorCache = [[NSMutableDictionary alloc] init];
        self.store = [SFSmartStore sharedStoreWithName:kDefaultSmartStoreName];
    }
    return self;
}


- (void)dealloc {
    self.store = nil;
    [super dealloc];
}

#pragma mark - Object bridging helpers


- (SFSoupCursor*)cursorByCursorId:(NSString*)cursorId
{
    return [_cursorCache objectForKey:cursorId];
}


- (void)closeCursorWithId:(NSString *)cursorId
{
    SFSoupCursor *cursor = [self cursorByCursorId:cursorId];
    if (nil != cursor) {
        [cursor close];
        [self.cursorCache removeObjectForKey:cursorId];
    } 
}

#pragma mark - SmartStore plugin methods

- (void)pgSoupExists:(NSMutableArray*)arguments withDict:(NSDictionary*)options
{
    //    NSDate *startTime = [NSDate date];
    NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgSoupExists" withArguments:arguments];
    NSString *soupName = [options nonNullObjectForKey:kSoupNameArg];
    
    BOOL exists = [self.store soupExists:soupName];
    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:exists];
    [self writeSuccessResultToJsRealm:result callbackId:callbackId];
    
//    NSLog(@"pgSoupExists took: %f", [startTime timeIntervalSinceNow]);
}

- (void)pgRegisterSoup:(NSMutableArray*)arguments withDict:(NSDictionary*)options
{
    //    NSDate *startTime = [NSDate date];
    NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgRegisterSoup" withArguments:arguments];
    NSString *soupName = [options nonNullObjectForKey:kSoupNameArg];
    NSArray *indexes = [options nonNullObjectForKey:kIndexesArg];
    
    BOOL regOk = [self.store registerSoup:soupName withIndexSpecs:indexes];
    if (regOk) {
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:soupName];
        [self writeSuccessResultToJsRealm:result callbackId:callbackId];
    } else {
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR ];
        [self writeErrorResultToJsRealm:result callbackId:callbackId];
    }
    
//    NSLog(@"pgRegisterSoup took: %f", [startTime timeIntervalSinceNow]);
}

- (void)pgRemoveSoup:(NSMutableArray*)arguments withDict:(NSDictionary*)options
{
//    NSDate *startTime = [NSDate date];
    NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgRemoveSoup" withArguments:arguments];
    NSString *soupName = [options nonNullObjectForKey:kSoupNameArg];
    
    [self.store removeSoup:soupName];
    
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK ];
    [self writeSuccessResultToJsRealm:result callbackId:callbackId];
    
//    NSLog(@"pgRemoveSoup took: %f", [startTime timeIntervalSinceNow]);
}

- (void)pgQuerySoup:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options
{
    NSDate *startTime = [NSDate date];
	NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgQuerySoup" withArguments:arguments];
    NSString *soupName = [options nonNullObjectForKey:kSoupNameArg];
    NSDictionary *querySpec = [options nonNullObjectForKey:kQuerySpecArg];
    
    SFSoupCursor *cursor =  [self.store querySoup:soupName withQuerySpec:querySpec];    
    NSLog(@"pgQuerySoup returning: %@",cursor);

    if (nil != cursor) {
        //cache this cursor for later paging
        [self.cursorCache setObject:cursor forKey:cursor.cursorId];
        [self writeSuccessDictToJsRealm:[cursor asDictionary] callbackId:callbackId];//TODO other error handling?
        NSLog(@"pgQuerySoup retrieved %d pages in %f",[cursor.totalPages integerValue], [startTime timeIntervalSinceNow]);
    } else {
        NSLog(@"No cursor for query: %@", querySpec);
        CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR ];
        [self writeErrorResultToJsRealm:result callbackId:callbackId];
    }
}

- (void)pgRetrieveSoupEntries:(NSMutableArray*)arguments withDict:(NSDictionary*)options
{
    NSDate *startTime = [NSDate date];
	NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgRetrieveSoupEntries" withArguments:arguments];
    NSString *soupName = [options nonNullObjectForKey:kSoupNameArg];
    NSArray *rawIds = [options nonNullObjectForKey:kEntryIdsArg];
    //make entry Ids unique
    NSSet *entryIdSet = [NSSet setWithArray:rawIds];
    NSArray *entryIds = [entryIdSet allObjects];
    
    NSArray *entries = [self.store retrieveEntries:entryIds fromSoup:soupName];
    [self writeSuccessArrayToJsRealm:entries callbackId:callbackId];
    
    NSLog(@"pgRetrieveSoupEntries in %f", [startTime timeIntervalSinceNow]);
}

- (void)pgUpsertSoupEntries:(NSMutableArray*)arguments withDict:(NSDictionary*)options
{
//    NSDate *startTime = [NSDate date];
	NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgUpsertSoupEntries" withArguments:arguments];
    NSString *soupName = [options nonNullObjectForKey:kSoupNameArg];
    NSArray *entries = [options nonNullObjectForKey:kEntriesArg];
    NSString *externalIdPath = [options nonNullObjectForKey:kExternalIdPathArg];
    
    NSError *error = nil;
    NSArray *resultEntries = [self.store upsertEntries:entries toSoup:soupName withExternalIdPath:externalIdPath error:&error];
    CDVPluginResult *result;
    if (nil != resultEntries) {
        //resultEntries
        [self writeSuccessArrayToJsRealm:resultEntries callbackId:callbackId];
    } else {
        if (error == nil) {
            result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR ];
        } else {
            result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[error localizedDescription]];
        }
        [self writeErrorResultToJsRealm:result callbackId:callbackId];
    }
    
//    NSLog(@"pgUpsertSoupEntries upserted %d entries in %f",[entries count], [startTime timeIntervalSinceNow]);
}

- (void)pgRemoveFromSoup:(NSMutableArray*)arguments withDict:(NSDictionary*)options
{
//    NSDate *startTime = [NSDate date];
	NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgRemoveFromSoup" withArguments:arguments];    
    NSString *soupName = [options nonNullObjectForKey:kSoupNameArg];
    NSArray *entryIds = [options nonNullObjectForKey:kEntryIdsArg];
    
    [self.store removeEntries:entryIds fromSoup:soupName];
    
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK ];
    [self writeSuccessResultToJsRealm:result callbackId:callbackId];
    
//    NSLog(@"pgRemoveFromSoup took: %f", [startTime timeIntervalSinceNow]);
    
}

- (void)pgCloseCursor:(NSMutableArray*)arguments withDict:(NSDictionary*)options
{
	NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgCloseCursor" withArguments:arguments];    
    NSString *cursorId = [options nonNullObjectForKey:kCursorIdArg];
    
    [self closeCursorWithId:cursorId];
    
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK ];
    [self writeSuccessResultToJsRealm:result callbackId:callbackId];
}

- (void)pgMoveCursorToPageIndex:(NSMutableArray*)arguments withDict:(NSDictionary*)options
{
    NSDate *startTime = [NSDate date];
	NSString* callbackId = [arguments pop];
    /* NSString* jsVersionStr = */[self popVersion:@"pgMoveCursorToPageIndex" withArguments:arguments];
    NSString *cursorId = [options nonNullObjectForKey:kCursorIdArg];
    NSNumber *newPageIndex = [options nonNullObjectForKey:kIndexArg];
    NSLog(@"pgMoveCursorToPageIndex: %@ [%d]",cursorId,[newPageIndex integerValue]);
    
    SFSoupCursor *cursor = [self cursorByCursorId:cursorId];
    [cursor setCurrentPageIndex:newPageIndex];
    
    [self writeSuccessDictToJsRealm:[cursor asDictionary] callbackId:callbackId];    
    
    NSLog(@"pgMoveCursorToPageIndex took: %f", [startTime timeIntervalSinceNow]);
}



@end
